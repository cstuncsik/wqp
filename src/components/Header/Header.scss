.Header {
  display: flex;
  padding: 20px;
  align-items: center;
  justify-content: space-between;
  box-sizing: border-box;
  grid-area: Header;
}
