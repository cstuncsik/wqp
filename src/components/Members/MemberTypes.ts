import { Dictionary } from 'helpers/types'

export interface Member {
  person_id: string
  firstname: string
  lastname: string
  title: string
  business_unit: string
  is_user: boolean
}

export interface MembersFCProps {
  members: Member[]
}

export interface MemberAccessLevel {
  id: 'read' | 'write' | 'admin'
  label: string
}

export interface MemberRoleSettings {
  role: 'customer' | 'employee' | 'manager'
  access_level: MemberAccessLevel['id']
}

export type MemberWithRole = Member & MemberRoleSettings

export interface MemberRole {
  id: MemberRoleSettings['role']
  label: string
  accessLevels: MemberAccessLevel['id'][]
}

export interface MemberRowFCProps {
  idx: number
  selectedMember: MemberWithRole
  selectableMembers: MemberWithRole[]
  removeMember: (idx: number) => void
  updateMember: (idx: number, data: MemberWithRole) => void
  switchMember: (idx: number, newMemberId: string) => void
  roles: Dictionary<MemberRole>
  accessLevels: Dictionary<MemberAccessLevel>
}

export type MemberResult = Partial<MemberWithRole>
