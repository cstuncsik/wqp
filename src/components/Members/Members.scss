.Members {
  display: grid;
  padding: 20px;
  grid-template-rows: 1fr auto auto;
  grid-row-gap: 20px;
  grid-area: Members;
}

.Members-data {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-auto-rows: min-content;
  grid-column-gap: 20px;
  grid-row-gap: 20px;
  align-items: center;
  border-top: 2px solid #ccc;
}

.Members-result {
  padding: 20px 0 0;
  border-top: 2px solid #ccc;

  pre {
    margin: 0;
    padding-top: 20px;
    color: #fff;
  }
}
