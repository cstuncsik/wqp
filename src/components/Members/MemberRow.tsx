import React, { FC, Fragment } from 'react'

import { MemberAccessLevel, MemberRole, MemberRoleSettings, MemberRowFCProps } from 'components/Members/MemberTypes'

export const MemberRow: FC<MemberRowFCProps> = ({
  idx,
  selectedMember,
  selectableMembers,
  removeMember,
  updateMember,
  switchMember,
  roles,
  accessLevels
}) => {
  return (
    <Fragment>
      <select
        name="member"
        disabled={selectableMembers.length === 1}
        value={selectedMember.person_id}
        onChange={({ target }) => {
          switchMember(idx, target.value)
        }}
      >
        {selectableMembers.map(option => (
          <option key={option.person_id} value={option.person_id}>
            {option.firstname} {option.lastname}
          </option>
        ))}
      </select>
      <select
        name="role"
        value={selectedMember.role}
        onChange={({ target }) => {
          selectedMember.role = target.value as MemberRoleSettings['role']
          selectedMember.access_level = roles[selectedMember.role].accessLevels[0]
          updateMember(idx, selectedMember)
        }}
      >
        {Object.keys(roles).map(key => {
          const role: MemberRole = roles[key]
          return (
            <option key={role.id} value={role.id}>
              {role.label}
            </option>
          )
        })}
      </select>
      <select
        name="accessLevel"
        disabled={roles[selectedMember.role].accessLevels.length === 1}
        value={selectedMember.access_level}
        onChange={({ target }) => {
          selectedMember.access_level = target.value as MemberAccessLevel['id']
          updateMember(idx, selectedMember)
        }}
      >
        {roles[selectedMember.role].accessLevels.map(accessLevel => (
          <option key={accessLevel} value={accessLevels[accessLevel].id}>
            {accessLevels[accessLevel].label}
          </option>
        ))}
      </select>
      <div>
        <button type="button" onClick={() => removeMember(idx)}>
          X
        </button>
      </div>
    </Fragment>
  )
}
