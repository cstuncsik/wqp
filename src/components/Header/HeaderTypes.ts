import { Member } from 'components/Members/MemberTypes'

export interface HeaderFCProps {
  me: Member | null
}
