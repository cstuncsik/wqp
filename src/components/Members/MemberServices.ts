import { AxiosPromise } from 'axios'
import { http } from 'services/http'
import { Member } from 'components/Members/MemberTypes'

export const getMe = (): AxiosPromise<Member> => http.get('/me')
export const getMembers = (): AxiosPromise<Member[]> => http.get('/members')
