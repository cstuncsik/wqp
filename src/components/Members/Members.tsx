import React, { FC, useEffect, useState, Fragment } from 'react'

import {
  Member,
  MemberAccessLevel,
  MemberRole,
  MemberRoleSettings,
  MembersFCProps,
  MemberWithRole,
  MemberResult
} from 'components/Members/MemberTypes'
import { Dictionary } from 'helpers/types'
import { MemberRow } from 'components/Members/MemberRow'
import { pick } from 'helpers/object'

import 'components/Members/Members.scss'

const roles: Dictionary<MemberRole> = {
  customer: {
    id: 'customer',
    label: 'Customer',
    accessLevels: ['read', 'write']
  },
  employee: {
    id: 'employee',
    label: 'Employee',
    accessLevels: ['write', 'admin']
  },
  manager: {
    id: 'manager',
    label: 'Manager',
    accessLevels: ['admin']
  }
}

const accessLevels: Dictionary<MemberAccessLevel> = {
  read: {
    id: 'read',
    label: 'Read'
  },
  write: {
    id: 'write',
    label: 'Write'
  },
  admin: {
    id: 'admin',
    label: 'Admin'
  }
}

const memberDefaultRoleProps: MemberRoleSettings = { role: 'customer', access_level: 'read' }

export const Members: FC<MembersFCProps> = ({ members }) => {
  const [selectedMembers, setSelectedMembers] = useState<MemberWithRole[]>([])
  const [availableMembers, setAvailableMembers] = useState<Member[]>([])
  const [membersResult, setMembersResult] = useState<{ members: MemberResult[] }>({ members: [] })

  const addMember = (): void => {
    const added: MemberWithRole = { ...availableMembers.shift()!, ...memberDefaultRoleProps }
    setSelectedMembers(selectedMembers.concat(added))
  }

  const removeMember = (idx: number): void => {
    const removed: MemberWithRole = selectedMembers.splice(idx, 1)[0]
    setSelectedMembers(selectedMembers)
    setAvailableMembers(availableMembers.concat(removed))
  }

  const updateMember = (idx: number, data: MemberWithRole): void => {
    selectedMembers[idx] = data
    setSelectedMembers([...selectedMembers])
  }

  const switchMember = (idx: number, newMemberId: string): void => {
    const newMemberIdx: number = availableMembers.findIndex(member => member.person_id === newMemberId)
    const removedAvailableMember: Member = availableMembers.splice(newMemberIdx, 1)[0]
    const added: MemberWithRole = { ...removedAvailableMember, ...memberDefaultRoleProps }
    const removed: MemberWithRole = selectedMembers.splice(idx, 1, added)[0]
    setAvailableMembers(availableMembers.concat(removed))
  }

  const createMembersResult = (): void => {
    setMembersResult({ members: selectedMembers.map(pick(['person_id', 'role', 'access_level'])) })
  }

  useEffect(() => {
    setAvailableMembers([...members])
  }, [members])

  return (
    <div className="Members" data-testid="members">
      {members.length > 0 ? (
        <Fragment>
          <div className="Members-data">
            <h2>Member</h2>
            <h2>Role</h2>
            <h2>Access Level</h2>
            <span>
              <button type="button" onClick={addMember} disabled={!availableMembers.length}>
                Add new member
              </button>
            </span>
            {selectedMembers.map((selectedMember, idx) => {
              const selectableMembers: MemberWithRole[] = availableMembers.reduce(
                (additionalSelectableMembers, member) =>
                  selectedMember.person_id !== member.person_id
                    ? additionalSelectableMembers.concat({ ...member, role: 'customer', access_level: 'read' })
                    : additionalSelectableMembers,
                [selectedMember]
              )
              return (
                <MemberRow
                  key={selectedMember.person_id}
                  {...{
                    idx,
                    selectedMember,
                    selectableMembers,
                    removeMember,
                    updateMember,
                    switchMember,
                    roles,
                    accessLevels
                  }}
                />
              )
            })}
          </div>
          <div className="Members-result">
            <button type="button" onClick={createMembersResult}>
              Save result
            </button>
            {membersResult.members.length > 0 && <pre>{JSON.stringify(membersResult, null, 2)}</pre>}
          </div>
        </Fragment>
      ) : (
        <h2>No available members</h2>
      )}
    </div>
  )
}
