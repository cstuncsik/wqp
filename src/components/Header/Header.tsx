import React, { FC } from 'react'

import { HeaderFCProps } from 'components/Header/HeaderTypes'

import 'components/Header/Header.scss'

export const Header: FC<HeaderFCProps> = ({ me }) => {
  return (
    <header className="App-header" data-testid="header">
      {me ? (
        <h1>
          {me?.firstname} {me?.lastname}
        </h1>
      ) : (
        <h1>No user data</h1>
      )}
      <button type="button">New user</button>
    </header>
  )
}
