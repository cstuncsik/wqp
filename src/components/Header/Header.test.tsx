import React from 'react'
import { render } from '@testing-library/react'

import { Header } from 'components/Header/Header'
import { Member } from 'components/Members/MemberTypes'

describe('Header component', () => {
  it('shows user name if given a valid user', () => {
    const me: Member = {
      person_id: '1b31592d-3d8a-42ff-9a8a-a373b10c2ade',
      firstname: 'Csaba',
      lastname: 'Tuncsik',
      title: 'Frontend developer',
      business_unit: 'Product',
      is_user: true
    }
    const { container } = render(<Header me={me} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('shows a message if not given a user', () => {
    const me = null
    const { container } = render(<Header me={me} />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
