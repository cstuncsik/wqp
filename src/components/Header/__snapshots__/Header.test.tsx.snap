// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component shows a message if not given a user 1`] = `
<header
  class="Header"
>
  <h1>
    No user data
  </h1>
</header>
`;

exports[`Header component shows user name if given a valid user 1`] = `
<header
  class="Header"
>
  <h1>
    Csaba
     
    Tuncsik
  </h1>
</header>
`;
